core/src/impl/Kokkos_Core.cpp

```diff
@@ -281,57 +310,92 @@ int Kokkos::Impl::get_ctest_gpu(const char* local_rank_str) {
+  if (env_visible_devices) {
+    std::stringstream ss(env_visible_devices);
+    for (int i; ss >> i;) {
+      visible_devices.push_back(i);
+      if (ss.peek() == ',') ss.ignore();
+    }
+  } else {
+    int num_devices =
+        settings.has_num_devices() ? settings.get_num_devices() : device_count;
+    for (int i = 0; i < num_devices; ++i) {
+      visible_devices.push_back(i);
+    }
+    if (settings.has_skip_device()) {
+      if (visible_devices.size() == 1 && settings.get_skip_device() == 0) {
+        Kokkos::abort(
+            "Error: skipping the only GPU available for execution.\n");
       }
+      visible_devices.erase(
+          std::remove(visible_devices.begin(), visible_devices.end(),
+                      settings.get_skip_device()),
+          visible_devices.end());
     }
-    // shift assignments over by one so no one is assigned to "skip_device"
-    if (use_gpu >= skip_device) ++use_gpu;
   }
-  return use_gpu;
+  if (visible_devices.empty()) {
+    Kokkos::abort("Error: no GPU available for execution.\n");
+  }
+  return visible_devices;
+}
+
+int Kokkos::Impl::get_gpu(const InitializationSettings& settings) {
+  std::vector<int> visible_devices =
+      get_visible_devices(settings, get_device_count());
+  // device_id is provided
+  if (settings.has_device_id()) {
+    return visible_devices[settings.get_device_id()];
+  }
+  // by default use the first GPU available for execution
+  // (neither device_id nor map_device_id_by are provided)
+  if (!settings.has_map_device_id_by()) {
+    return visible_devices[0];
+  }
+  // map_device_id provided
+  // either random or round-robin assignement based on local MPI rank
+  if (!is_valid_map_device_id_by(settings.get_map_device_id_by())) {
+    std::cerr << "Warning: unrecognized map_device_id_by setting \""
+              << settings.get_map_device_id_by() << "\" ignored."
+              << " Raised by Kokkos::initialize(int argc, char* argv[])."
+              << std::endl;
+    return visible_devices[0];
+  }
+
+  if (settings.get_map_device_id_by() == "random") {
+    std::default_random_engine gen(get_process_id());
+    std::uniform_int_distribution<int> distribution(0,
+                                                    visible_devices.size() - 1);
+    return visible_devices[distribution(gen)];
+  }
+
+  if (settings.get_map_device_id_by() != "mpi_rank") {
+    Kokkos::abort("implementation bug");
+  }
+
+  auto const* local_rank_str =
+      std::getenv("OMPI_COMM_WORLD_LOCAL_RANK");  // OpenMPI
+  if (!local_rank_str)
+    local_rank_str = std::getenv("MV2_COMM_WORLD_LOCAL_RANK");  // MVAPICH2
+  if (!local_rank_str) local_rank_str = std::getenv("SLURM_LOCALID");  // SLURM
+
+  if (!local_rank_str) {
+    std::cerr << "Warning: unable to detect local MPI rank."
+              << " Raised by Kokkos::initialize(int argc, char* argv[])."
+              << std::endl;
+    return visible_devices[0];
```

Comment:
I would like to give a hint to the user, that we default to device 0 in this case